### PR TITLE
Diagnostics that hints about use of std.array.array upon implicit conversion in assignments

### PR DESCRIPTION
**Pull request by nordlow**
This adds logic that hints the **new** developer on when to use `std.array.array` to force evaluation of a lazy range and write it into an array when trying to assign a range to a D array as in

``` D
    int[] x = [1, 2, 3].map!"a*2";
```

gives error:

```
cannot implicitly convert expression (map([1, 2, 3])) of type MapResult!(unaryFun, int[]) to int[], use .array (in std.array) to force allocation
```

Assignment to non-arrays such as

``` D
    int x = [1, 2, 3].map!"a*2";
```

has unchanged diagnostics.

For background discussion see

http://forum.dlang.org/thread/xyvznufauehijucwwjav@forum.dlang.org?page=1

This is in an early stage. Feedback, please!

### CODE REVIEW

File: src/cast.c [hunk 1]
```diff
@@ -100,8 +101,28 @@ Expression *implicitCastTo(Expression *e, Scope *sc, Type *t)
...
+                char hint[128];
+                snprintf(hint, sizeof(hint), "");
+                if (t->isTypeDArray())
+                {
+                    if (TypeStruct* typeStruct = e->type->isTypeStruct())
+                    {
+                        if (typeStruct->sym->isInputRange())
+                        {
+                            snprintf(hint, sizeof(hint), ", suffix Range with std.array.array");
+                        }
```
Comment (schuetzm) [inline src/cast.c]: I suggest to rephrase that, as it can be easily misread (what is a "suffix range"?).  How about, "use std.array.array to force conversion".

File: src/class.c [hunk 2]
```diff
@@ -1205,6 +1205,17 @@ void ClassDeclaration::addLocalClass(ClassDeclarations *aclasses)
...
+ * Return true if class is an InputRange.
+ * This is defined as having the members: empty, front, popFront.
+ */
+bool ClassDeclaration::isInputRange()
+{
+    return (this->search(Loc(), Id::Ffront) &&
+            this->search(Loc(), Id::FpopFront) &&
+            this->search(Loc(), Id::Fempty));
+}
+
```
Comment (schuetzm) [inline src/class.c]: Together with the `foreach` handling code in https://github.com/D-Programming-Language/dmd/blob/master/src/statement.c#L2094 (which btw tries range iteration if it finds `front`, but will later fail if `empty` and `popFront()` are not defined), there are now three places to recognize an input range. Can they be unified?

Comment (nordlow) [inline src/class.c]: Absolutely :) I can fix foreach if you want to. Do you want me defined predicates for other ranges aswell?

Comment (nordlow) [inline src/class.c]: Where's the third place input ranges are recognized?

File: src/cast.c (hunk 1 above)
Comment (nordlow) [inline src/cast.c]: Fixed.

File: src/class.c (hunk 2 above)
Comment (schuetzm) [inline src/class.c]: Not sure about foreach, maybe it is intentional? The current behaviour has the effect of detecting ill-defined input ranges by erroring, instead of falling back to (trying) opApply (which will likely not exist in most cases). In any case, there should be only one notion of input range in DMD.

Other range types can be added on demand later, maybe they will be needed for better diagnostics for other range related errors?

Comment (schuetzm) [inline src/class.c]: ClassDeclaration + StructDeclaration + foreach == 3

File: src/cast.c (hunk 1 above)
Comment (nordlow) [inline src/cast.c]: Or what about

```
, use .array in std.array to force conversion
```

?

File: src/class.c (hunk 2 above)
Comment (nordlow) [inline src/class.c]: Got it.

Comment (nordlow) [inline src/class.c]: There is probably other diagnostics that may find other range predicates useful. But why not add them all at once in this PR?

**Comment by nordlow** [comment]
BTW: Can/Should Ranges be classes aswell as structs in D?

**Comment by nordlow** [comment]
BTW: Shouldn't `ScopeDsymbol::search()` be declared virtual because of the subclass member `StructDeclaration:;search()` having the same signature?!

File: src/cast.c [hunk 3]
```diff
@@ -100,8 +100,28 @@ Expression *implicitCastTo(Expression *e, Scope *sc, Type *t)
                 //printf("type %p ty %d deco %p\n", type, type->ty, type->deco);
                 //type = type->semantic(loc, sc);
                 //printf("type %s t %s\n", type->deco, t->deco);
-                e->error("cannot implicitly convert expression (%s) of type %s to %s",
-                    e->toChars(), e->type->toChars(), t->toChars());
+                char hint[128];
```
Comment (yebblies) [inline src/cast.c]: There is no need to use a buffer to store string literals, just use a `const char*`

**Comment by yebblies** [comment]
Please remove all the new functions you've defined that aren't used.

File: src/struct.c [hunk 4]
```diff
@@ -631,6 +631,52 @@ Dsymbol *AggregateDeclaration::searchCtor()
     return s;
 }
 
+/***************************************
+ * Return true if aggregate is an InputRange.
+ */
+bool AggregateDeclaration::isInputRange()
+{
+    return (this->search(Loc(), Id::Ffront) &&
```
Comment (JakobOvrum) [inline src/struct.c]: Doesn't this simply check if the aggregate has a member `front`? Simply having members named `front`, `popFront` and `empty` does not make an input range: http://dlang.org/phobos/std_range#isInputRange

File: src/cast.c [hunk 5]
```diff
@@ -100,8 +100,28 @@ Expression *implicitCastTo(Expression *e, Scope *sc, Type *t)
...
+                sprintf(hint, "");
+                if (t->isTypeDArray())
+                {
+                    if (TypeStruct* typeStruct = e->type->isTypeStruct())
```
Comment (JakobOvrum) [inline src/cast.c]: These two paths are almost the same, surely they can be merged into a single neater path that doesn't duplicate the hint message?

File: src/struct.c (hunk 4 above)
Comment (nordlow) [inline src/struct.c]: Do I have to check

- that the return type `front()` is assignable to the element of the D darray and
- that empty returns `bool` and
- that `popFront()` takes no arguments and returns `void`?

Comment (nordlow) [inline src/struct.c]: Do you mean that is should all range predicates except `isInputRange`?

Comment (JakobOvrum) [inline src/struct.c]: `front` and `empty` do not have to be functions, but indeed when they are functions, they must be nullary, and `front` must be of a non-void type. Strictly speaking, `empty` does not have to be of `bool` type but of a type castable to `bool`. `popFront` must be a nullary function.

Additionally, these properties do not have to be implemented as members, they can be implemented as free-standing property functions, which satisfy the concept through UFCS.

I don't know how foreach deals with it, but I'm guessing it uses error suppression, much like the `isInputRange` checker template itself.

Comment (nordlow) [inline src/struct.c]: Should I implement checks for this?

I'm guessing I should use `Scope::search()` when searching for free-standing property functions right?

For every such free function `f` how I do check that it's callable as `f(e)`?

File: src/cast.c [hunk 6]
```diff
@@ -100,8 +100,32 @@ Expression *implicitCastTo(Expression *e, Scope *sc, Type *t)
                 //printf("type %p ty %d deco %p\n", type, type->ty, type->deco);
                 //type = type->semantic(loc, sc);
                 //printf("type %s t %s\n", type->deco, t->deco);
-                e->error("cannot implicitly convert expression (%s) of type %s to %s",
-                    e->toChars(), e->type->toChars(), t->toChars());
+                const char *hint = "";
+                if (t->isTypeDArray())
+                {
+                    bool doHint = false;
+                    if (TypeStruct* typeStruct = e->type->isTypeStruct())
```
Comment (9rnsr) [inline src/cast.c]: Instead you can use `isAggregate` function.

```d
AggregateDeclaration *ad = isAggregate(e->type);
if (ad && ad->isInputRange())
    hint = "..."
```

File: src/aggregate.h [hunk 7]
```diff
@@ -128,6 +128,8 @@ class AggregateDeclaration : public ScopeDsymbol
 
     AggregateDeclaration *isAggregateDeclaration() { return this; }
     void accept(Visitor *v) { v->visit(this); }
+
+    virtual bool isInputRange();
```
Comment (9rnsr) [inline src/aggregate.h]: `virtual` is unnecessary.

File: src/mtype.h [hunk 8]
```diff
@@ -268,8 +271,15 @@ class Type : public RootObject
     virtual bool isscope();
     virtual bool isString();
     virtual bool isAssignable();
+
+    virtual TypeStruct* isTypeStruct() { return NULL; }
+    virtual TypeClass* isTypeClass() { return NULL; }
+    virtual bool isTypeAggregate() { return this->isTypeStruct() || this->isTypeClass(); }
+    virtual TypeDArray* isTypeDArray() { return NULL; }
```
Comment (9rnsr) [inline src/mtype.h]: All `isTypeXXX` functions are redundant. Please use `type->ty == Txxxx` instead.

**Comment by 9rnsr** [comment]
I'm really not sure that dmd should support such a hint for `std.array.array` function.

File: src/mtype.h (hunk 8 above)
Comment (nordlow) [inline src/mtype.h]: But usage of `isTypeStruct()´ in if-expression is easier to read. If you're afraid of virtual functions being too slow why not defined and use (inline) wrappers around type->ty == Txxxx plus cast.

Please consider this. I'm trying to make the DMD code base easier to understand and work with, especially for newcomers (like me).

**Comment by nordlow** [comment]
Is it worse than hinting about `writeln`?

**Comment by JakobOvrum** [comment]
Using ranges correctly requires an understanding of their lazily-evaluated semantics. I'm a little worried that this hint would lead users to create superfluous arrays when what they should have done is use the range lazily.

**Comment by nordlow** [comment]
Yes, I'm aware of that risk.

File: src/mtype.h (hunk 8 above)
Comment (9rnsr) [inline src/mtype.h]: I like the inline style on if-statement, but I'll never wish adding such `isTypeXXX()` functions.

Because it doesn't increase code readability. Today, all dmd developers should learn the relation between `Txxx` enums and `TypeXXX` classes. So such the functions will just increase the code style confusion.

File: src/struct.c (hunk 4 above)
Comment (nordlow) [inline src/struct.c]: Unless we replace them *all* with `isXType` or `dynamic_cast` that is...I'm simply wishing the DMD code base to one day become as clean in readability as Clang is.

File: src/mtype.h (hunk 8 above)
Comment (nordlow) [inline src/mtype.h]: Unless we replace them *all* with `isXType` or `dynamic_cast` that is...I'm simply wishing the DMD code base to one day become as clean in readability as Clang is.

Comment (nordlow) [inline src/mtype.h]: Further there are already usages of this pattern in `class Dsymbol`.

Are you saying that these should be removed?

Comment (yebblies) [inline src/mtype.h]: Readability is subjective.  Sometimes the isXXX methods are nicer, sometimes they're not.  Adding a new redundant way to do something is not usually a good idea.  The worst of all is those changes inevitable result in discussions like this instead of useful discussion about the actual changes.

> Further there are already usages of this pattern in class Dsymbol.

Yes, `Dsymbol` uses this pattern, because `Dsymbol` doesn't have `Txxx`.  `Expression` uses `TOKxxx`.

Comment (9rnsr) [inline src/mtype.h]: `Dsymbol` and its derived classes don't have a field similar to `Type::ty` or `Expression::op`. So it's impossible.

When you think we should prefer `isXXX()` style in all places, I think you would need to propose removing `Type::ty` and `Expression::op`.

But I doubt that it will increase dmd code quality. At least `type->ty == Txxx` works enough.

Comment (nordlow) [inline src/mtype.h]: Not quite... The user may make mistakes such as

```C
if (e->op == TOKtemplate)
{
    NonTemplateExp* ntx = (NonTemplateExp*)e;
}
```

which is not caught by the compiler.

Making mistakes with my alternative

if (TemplateExp* te = e->isTemplateExp())
{
    ...
}
```

will be caught by the compiler provided that isTemplateExp is implemented correctly which of course has to be dealt with but that's only in one place.

Comment (yebblies) [inline src/mtype.h]: If you really want to make a case for adding these methods, do it in another pull request.

Comment (nordlow) [inline src/mtype.h]: Ok. Great. Thanks. Do you think we should add them all at once or one per TOK enumerator value?

Comment (yebblies) [inline src/mtype.h]: I don't think we should add them at all.

**Comment by schuetzm** [comment]
That can be easily mitigated: "Use std.array.array to allocate an array with the range's elements."

**Comment by nordlow** [comment]
So...should I change my hint message to be more verbose

```
use std.array.array to allocate an array with the Range's elements
```

then?

**Comment by nordlow** [comment]
What about being more specific about Range type in hint message?

**Comment by nordlow** [comment]
Ping. What's next?

**Comment by nordlow** [comment]
Updated after D conversion.

Code example above now errors as:

```
cannot implicitly convert expression (map([1, 2, 3])) of type MapResult!(unaryFun, int[]) to int[], use .array (in std.array) to force allocation
```

File: src/aggregate.d [hunk 9]
```diff
@@ -639,6 +639,16 @@ extern (C++) abstract class AggregateDeclaration : ScopeDsymbol
         return type;
     }
 
+    /****************************************
+     * Returns `true` if `this` aggregate has an `InputRange` interface
+     */
+    final bool isInputRange()
```
Comment (JackStouffer) [inline src/aggregate.d]: As `foreach` already tests for ranges, I would have to guess this test exists somewhere in dmd already

File: src/aggregate.h [hunk 10]
```diff
@@ -313,6 +313,7 @@ class ClassDeclaration : public AggregateDeclaration
     Symbol *vtblsym;
 
     ClassDeclaration *isClassDeclaration() { return (ClassDeclaration *)this; }
+
```
Comment (JackStouffer) [inline src/aggregate.h]: Superfluous

**Comment by JackStouffer** [comment]
> I guess I should add a test for the new diagnostics message as well, right?

Yes

File: src/aggregate.d (hunk 9 above)
Comment (mathias-lang-sociomantic) [inline src/aggregate.d]: [I like your optimism](https://github.com/dlang/dmd/blob/38c1bad2f41ad8e6bc7c08deae26e30c9c7d7704/src/statementsem.d#L999-L1039)

Comment (JackStouffer) [inline src/aggregate.d]: wow

File: src/aggregate.h (hunk 10 above)
Comment (nordlow) [inline src/aggregate.h]: Removed.

File: src/aggregate.d [hunk 11]
```diff
@@ -639,6 +639,16 @@ extern (C++) abstract class AggregateDeclaration : ScopeDsymbol
...
+    {
+        return (search(Loc(), Id.Ffront) &&
+                search(Loc(), Id.FpopFront) &&
+                search(Loc(), Id.Fempty));
+    }
+
```
Comment (WalterBright) [inline src/aggregate.d]: While a good idea to encapsulate this sort of thing, if this returns true the caller will have to call search() again to find it.

File: src/dcast.d [hunk 12]
```diff
@@ -97,8 +97,17 @@ extern (C++) Expression implicitCastTo(Expression e, Scope* sc, Type t)
...
-                    e.error("cannot implicitly convert expression (%s) of type %s to %s",
-                        e.toChars(), e.type.toChars(), t.toChars());
+                    immutable(char)* arrayHint = "";
+                    if (t.ty == Tarray)
+                    {
+                        auto ad = isAggregate(e.type);
+                        if (ad && ad.isInputRange())
+                        {
+                            arrayHint = ", use .array (in std.array) to force allocation of range elements";
+                        }
```
Comment (WalterBright) [inline src/dcast.d]: I just don't think this is a good idea for such messages, it is too hard to guess what the user might be trying to accomplish. Guess wrong, and things can be very confusing.

File: src/mtype.h [hunk 13]
```diff
@@ -492,6 +492,8 @@ class TypeDArray : public TypeArray
     bool hasPointers() /*const*/;
 
     void accept(Visitor *v) { v->visit(this); }
+
+    TypeDArray* isTypeDArray() { return this; }
```
Comment (WalterBright) [inline src/mtype.h]: These functions are never called.

**Comment by WalterBright** [comment]
I'm going to close it for now. It needs some significant re-thinking. Reopen when ready.

File: src/dcast.d (hunk 12 above)
Comment (nordlow) [inline src/dcast.d]: If so, does @WalterBright or anyone else have a suggestion on a different approach?
